scapy/contrib/automotive/someip.py

```diff
@@ -225,3 +206,326 @@ def _bind_someip_layers():
 
 
 _bind_someip_layers()
+
+
+class _SDPacketBase(Packet):
+    """ base class to be used among all SD Packet definitions."""
+    # use this dictionary to set default values for desired fields (mostly on
+    # subclasses where not all fields are defined locally)
+    # - key : field_name, value : desired value
+    # - it will be used from 'init_fields' function, upon packet initialization
+    #
+    # example : _defaults =
+    #  {'field_1_name':field_1_value,'field_2_name':field_2_value}
+    _defaults = {}
+
+    def _set_defaults(self):
+        for key in self._defaults:
+            try:
+                self.get_field(key)
+            except KeyError:
+                pass
+            else:
+                self.setfieldval(key, self._defaults[key])
+
+    def init_fields(self):
+        super(_SDPacketBase, self).init_fields()
+        self._set_defaults()
+
+
+# SD ENTRY
+#  - Service
+#  - EventGroup
+class _SDEntry(_SDPacketBase):
+    TYPE_FMT = ">B"
+    TYPE_PAYLOAD_I = 0
+    TYPE_SRV_FINDSERVICE = 0x00
+    TYPE_SRV_OFFERSERVICE = 0x01
+    TYPE_SRV = (TYPE_SRV_FINDSERVICE, TYPE_SRV_OFFERSERVICE)
+    TYPE_EVTGRP_SUBSCRIBE = 0x06
+    TYPE_EVTGRP_SUBSCRIBE_ACK = 0x07
+    TYPE_EVTGRP = (TYPE_EVTGRP_SUBSCRIBE, TYPE_EVTGRP_SUBSCRIBE_ACK)
+    OVERALL_LEN = 16
+
+    fields_desc = [
+        ByteField("type", 0),
+        ByteField("index_1", 0),
+        ByteField("index_2", 0),
+        BitField("n_opt_1", 0, 4),
+        BitField("n_opt_2", 0, 4),
+        ShortField("srv_id", 0),
+        ShortField("inst_id", 0),
+        ByteField("major_ver", 0),
+        X3BytesField("ttl", 0)
+    ]
+
+    def guess_payload_class(self, payload):
+        pl_type = orb(payload[_SDEntry.TYPE_PAYLOAD_I])
+
+        if pl_type in _SDEntry.TYPE_SRV:
+            return SDEntry_Service
+        elif pl_type in _SDEntry.TYPE_EVTGRP:
+            return SDEntry_EventGroup
+
+
+class SDEntry_Service(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_SRV_FINDSERVICE}
+
+    name = "Service Entry"
+    fields_desc = [
+        _SDEntry,
+        IntField("minor_ver", 0)
+    ]
+
+
+class SDEntry_EventGroup(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_EVTGRP_SUBSCRIBE}
+
+    name = "Eventgroup Entry"
+    fields_desc = [
+        _SDEntry,
+        BitField("res", 0, 12),
+        BitField("cnt", 0, 4),
+        ShortField("eventgroup_id", 0)
+    ]
+
+
+# SD Option
+#  - Configuration
```

Comment:
There are two white-spaces here and below, do you want to change that?

Reply:
sure :)